File:  Maximum Subarray/ Maximum Subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        
        max_so_far = nums[0]
        curr_value = nums[0]
        for i in range(1, len(nums)):
            curr_value = max(nums[i], curr_value + nums[i])
            max_so_far = max(max_so_far, curr_value)
        return max_so_far
        

File:  Maximum Subarray/ Maximum Subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        
        int maxSum = nums[0], currentSum = 0;
        
        for (int& num : nums) {
            currentSum += num;
            
            if ( currentSum > maxSum )
                maxSum = currentSum;
            
            if ( currentSum < 0 )
                currentSum = 0;
        }
        
        return maxSum;
    }
};